File: dsa/2-bigO/quadratic-time-O(n^2)/quadraticTime.js
import { performance } from 'perf_hooks';

const numbers = Array.from({ length: 100 }, (_, i) => i + 1)

function findPairs(arr) {
    // O(n^2)
    for (let i = 0; i < arr.length; i++) {
        for (let j = i + 1; j < arr.length; j++) {
            console.log(`Pair: ${arr[i]}, ${arr[j]}`)
        }
    }
}

// Benchmarking the function
const iterations = 1;
let start = performance.now();
for (let i = 0; i < iterations; i++) {
    findPairs(numbers);
}
let end = performance.now();
console.log(`findPairs O(n^2) time: ${(end - start).toFixed(6)} ms for ${iterations} iteration(s)`);

File: dsa/2-bigO/quadratic-time-O(n^2)/quadratic-time.py
import timeit

numbers = list(range(1, 101))


def find_pairs(arr):
    # O(n^2)
    for i in range(len(arr)):
        for j in range(i + 1, len(arr)):
            print(f"Pair: {arr[i]}, {arr[j]}")


# Benchmarking the function
iterations = 1
time_on2 = timeit.timeit("find_pairs(numbers)", globals=globals(), number=iterations)
print(f"find_pairs O(n^2) time: {time_on2 * 1000:.3f} ms for {iterations} iterations")
